sql: Make JSON parsing faster.

miretskiy opened the pull request:
`tree.ParseJSON` is slow, and inefficient, primary due to its multi step
parsing implementation, where first, the input JSON string is decoded using 
Go standard `encoding/json` library, and then the Go native data type is 
converted to `tree.JSON`.

There are many inefficiencies in this process including reliance on reflection, 
unnecessary allocations (map[string]interface{}) and many more.  
The reality is that `encoding/json` does too many things that we do not need 
(i.e ability to decode into `struct{}`) and that hurts performance.

This PR introduces a new fast json parser implementation,
to directly convert input string into the `tree.JSON` representation. 
There are no intermediate steps.  As we process input string, 
we build the final `tree.JSON` representation.
This implementation does not use reflection, and, in general,
is substantially faster, and more memory efficient across all benchmarks.

JSON parser is built on top of somewhat heavily modified fork of `pkg/json` package.  
That package, while incomplete, provided an excellent starting point.
Namely its `Scanner` and `Decoder` implementations provided a blueprint 
on how to implement the higher level logic around direct conversion into `tree.JSON`.
The techniques demonstrated in this package
(https://dave.cheney.net/high-performance-json.html) were used throughout 
this package in order to build high performance state machine to convert 
strings into JSONb. This package was modified, somewhat heavily, to include 
required, but missing parts, including string un-escaping, as well as handling of UTF characters.

New implementation should be a drop in replacement to our use of `encoder/json` in `tree.ParseJSON`.
Existing tests were modified to test both implementation.  Extensive improvement were 
made to random JSON generation library so that the tests are more complete and thorough.

Performance of this new library substantially exceeds Go standard implementation
for common cases -- namely, the case when input strings do not contain escape sequences. 
The new implementation consistently requires fewer (and smaller) allocations, 
while at the same time being faster both in terms of time per op, as well as raw throughput. 
The benchmark comparing large input, with strings fields up to 1/2MB in size, 
the difference in throughput was over 8 *times*.

Handling escaped, or UTF encoded data in JSON stream, is an expensive proposition: 
as soon as the data is known to contain escape characters `\`, additional memory allocations 
and copies are needed.  This hurts performance.  While the new implementation is still faster, 
and more memory efficient, the differences are not as pronounced (only 75% faster when
encoding heavily escaped, almost binary data).

The final benefit of the new implementation is that we may choose (in the future) to no longer
sort object keys upon construction.  This capability is exposed via a
`jsonb.WithUnorderedObjectKeys()` option to `tree.ParseJSON`. 
The standard Go implementation had to sort the keys because otherwise, 
the same input will not produce round trip-able output because of non-deterministic 
order of map traversal in Go.  This is no longer applicable to this implementation 
since the objects are constructed in a well defined ordered. This option, by itself, results
in a modest 5-7% improvement in speed (and about 10% reduction in allocations).
While these improvements are not that significant, more possibilities exists to push the
envelope, perhaps by supporting some custom parsing options.

Fixes #89998 

Release note (performance improvement): significantly faster JSON parsing.  
This should help with any workloads that insert large amounts of JSON data, including IMPORT workloads.

Commit:
sql: Make JSON parsing faster.

`tree.ParseJSON` is slow, and inefficient, primary due
to its multi step parsing implementation, where first,
the input JSON string is decoded using Go standard
`encoding/json` library, and then the Go native data
type is converted to `tree.JSON`.

There are many inefficiencies in this process including
reliance on reflection, unnecessary allocations (map[string]interface{})
and many more.  The reality is that `encoding/json` does
too many things that we do not need (i.e ability to decode
into `struct{}`) and that hurts performance.

This PR introduces a new fast json parser implementation,
to directly convert input string into the `tree.JSON` representation.
There are no intermediate steps.  As we process input string,
we build the final `tree.JSON` representation.
This implementation does not use reflection, and, in general,
is substantially faster, and more memory efficient across all benchmarks.

JSON parser is built on top of somewhat heavily modified
fork of `pkg/json` package.  That package, while incomplete,
provided an excellent starting point.
Namely its `Scanner` and `Decoder` implementations provided
a blueprint on how to implement the higher level logic
around direct conversion into `tree.JSON`.
The techniques demonstrated in this package
(https://dave.cheney.net/high-performance-json.html)
were used throughout this package in order to build
high performance state machine to convert strings into JSONb.
This package was modified, somewhat heavily, to include required,
but missing parts, including string un-escaping, as well as
handling of UTF characters.

New JSON parser implementation should be a drop in replacement
to our use of `encoder/json` in `tree.ParseJSON`.
Existing tests were modified to test both implementation.
Extensive improvement were made to random JSON generation
library so that the tests are more complete and thorough.

Performance of this new library substantially exceeds
Go standard implementation for common cases -- namely, the case
when input strings do not contain escape sequences.
The new implementation consistently requires fewer
(and smaller) allocations, while at the same time
being faster both in terms of time per op, as well as raw throughput.
The benchmark comparing large input, with strings fields up
to 1/2MB in size, the difference in throughput was over 6 *times*
(~139MB/s for Go standard implementation, vs 862MB/sec on a
linux AMD machine `n2d-standard-32`; while on faster Mac CPU,
single core benchmark was able to push 1440MB/sec).

Handling escaped, or UTF encoded data in JSON stream, is an
expensive proposition: as soon as the data is known to contain
escape characters `\`, additional memory allocations and copies
are needed.  This hurts performance.  While the new implementation
is still faster, and more memory efficient, the differences are
not as pronounced.

The final benefit of the new implementation is that we may
choose (in the future) to no longer sort object keys upon
construction.  This capability is exposed via a
`jsonb.WithUnorderedObjectKeys()` option to `tree.ParseJSON`.
The standard Go implementation had to sort the keys because
otherwise, the same input will not produce round trip-able
output because of non-deterministic order of map traversal in
Go.  This is no longer applicable to new implementation since
the objects are constructed in a well defined ordered.
Additional performance tweaking options may be added later.

Release note (performance improvement): significantly faster
JSON parsing.  This should help with any workloads that insert
large amounts of JSON data, including IMPORT workloads.

miretskiy commented:
GoStandard vs Lexer implementation benchmarks are below.
The percentage part in the benchmark name indicates the amount of data
that required escaping: with 0% being the happy case, 3% modeling the case
when the data has modest amount of escaped data (few `"`, few utf characters, etc),
and a 40% simulating binary (or UTF8 heavy data).
The size portion indicates the maximum size of the strings inside generated
(random) JSON inputs.

While the speed impact is the most striking (particularly in the no-escape case),
the lexer implementation, across all benchmarks, has a much nicer and predictable
allocation behavior.

```
name                               old time/op    new time/op     delta
ParseJSON/escape=0.00%/128-32        5.50µs ± 5%     2.88µs ± 4%   -47.54%  (p=0.000 n=9+10)
ParseJSON/escape=0.00%/512-32        7.31µs ± 6%     3.19µs ± 1%   -56.39%  (p=0.000 n=10+8)
ParseJSON/escape=0.00%/4096-32       26.5µs ± 1%      7.2µs ± 2%   -72.76%  (p=0.000 n=8+9)
ParseJSON/escape=0.00%/16384-32      85.1µs ± 1%     18.3µs ± 1%   -78.49%  (p=0.000 n=9+10)
ParseJSON/escape=0.00%/65536-32       312µs ± 1%       62µs ± 2%   -80.12%  (p=0.000 n=9+10)
ParseJSON/escape=0.00%/262144-32     1.35ms ± 1%     0.23ms ± 0%   -82.98%  (p=0.000 n=8+9)
ParseJSON/escape=0.00%/524288-32     2.62ms ± 1%     0.43ms ± 3%   -83.66%  (p=0.000 n=9+9)
ParseJSON/escape=3.00%/128-32        5.83µs ± 3%     3.38µs ± 1%   -42.00%  (p=0.000 n=9+9)
ParseJSON/escape=3.00%/512-32        8.63µs ± 0%     5.58µs ± 0%   -35.33%  (p=0.000 n=7+8)
ParseJSON/escape=3.00%/4096-32       34.0µs ± 1%     24.9µs ± 1%   -26.72%  (p=0.000 n=9+8)
ParseJSON/escape=3.00%/16384-32       110µs ± 2%       88µs ± 1%   -20.53%  (p=0.000 n=9+10)
ParseJSON/escape=3.00%/65536-32       416µs ± 1%      342µs ± 0%   -17.87%  (p=0.000 n=9+9)
ParseJSON/escape=3.00%/262144-32     1.76ms ± 7%     1.38ms ± 1%   -21.86%  (p=0.000 n=8+9)
ParseJSON/escape=3.00%/524288-32     3.44ms ± 5%     2.79ms ± 1%   -18.79%  (p=0.000 n=10+9)
ParseJSON/escape=40.00%/128-32       6.52µs ± 3%     3.75µs ± 1%   -42.46%  (p=0.000 n=10+9)
ParseJSON/escape=40.00%/512-32       11.7µs ± 1%      6.7µs ± 1%   -42.54%  (p=0.000 n=8+8)
ParseJSON/escape=40.00%/4096-32      55.6µs ± 1%     32.1µs ± 0%   -42.31%  (p=0.000 n=10+8)
ParseJSON/escape=40.00%/16384-32      204µs ± 1%      117µs ± 1%   -42.43%  (p=0.000 n=10+9)
ParseJSON/escape=40.00%/65536-32      831µs ± 1%      469µs ± 0%   -43.53%  (p=0.000 n=10+8)
ParseJSON/escape=40.00%/262144-32    3.18ms ± 0%     1.83ms ± 1%   -42.28%  (p=0.000 n=7+10)
ParseJSON/escape=40.00%/524288-32    6.77ms ± 1%     3.60ms ± 0%   -46.78%  (p=0.000 n=10+10)

name                               old speed      new speed       delta
ParseJSON/escape=0.00%/128-32      30.6MB/s ± 4%   58.3MB/s ± 4%   +90.62%  (p=0.000 n=9+10)
ParseJSON/escape=0.00%/512-32      58.4MB/s ± 6%  133.7MB/s ± 1%  +129.04%  (p=0.000 n=10+8)
ParseJSON/escape=0.00%/4096-32      110MB/s ± 1%    405MB/s ± 2%  +267.14%  (p=0.000 n=8+9)
ParseJSON/escape=0.00%/16384-32     133MB/s ± 1%    618MB/s ± 1%  +364.94%  (p=0.000 n=9+10)
ParseJSON/escape=0.00%/65536-32     144MB/s ± 1%    723MB/s ± 2%  +402.76%  (p=0.000 n=9+10)
ParseJSON/escape=0.00%/262144-32    138MB/s ± 1%    810MB/s ± 0%  +488.64%  (p=0.000 n=8+9)
ParseJSON/escape=0.00%/524288-32    139MB/s ± 1%    862MB/s ± 3%  +520.11%  (p=0.000 n=9+9)
ParseJSON/escape=3.00%/128-32      29.5MB/s ± 3%   50.8MB/s ± 1%   +72.18%  (p=0.000 n=9+10)
ParseJSON/escape=3.00%/512-32      52.4MB/s ± 0%   81.0MB/s ± 0%   +54.64%  (p=0.000 n=7+8)
ParseJSON/escape=3.00%/4096-32     90.0MB/s ± 1%  122.9MB/s ± 1%   +36.48%  (p=0.000 n=9+8)
ParseJSON/escape=3.00%/16384-32     107MB/s ± 2%    135MB/s ± 1%   +25.74%  (p=0.000 n=9+10)
ParseJSON/escape=3.00%/65536-32     114MB/s ± 1%    139MB/s ± 0%   +21.74%  (p=0.000 n=9+9)
ParseJSON/escape=3.00%/262144-32    108MB/s ± 5%    139MB/s ± 1%   +28.12%  (p=0.000 n=7+9)
ParseJSON/escape=3.00%/524288-32    111MB/s ± 5%    138MB/s ± 1%   +24.48%  (p=0.000 n=10+9)
ParseJSON/escape=40.00%/128-32     29.9MB/s ± 3%   52.0MB/s ± 1%   +73.77%  (p=0.000 n=10+9)
ParseJSON/escape=40.00%/512-32     47.4MB/s ± 1%   82.5MB/s ± 1%   +74.05%  (p=0.000 n=8+8)
ParseJSON/escape=40.00%/4096-32    68.5MB/s ± 1%  118.8MB/s ± 0%   +73.33%  (p=0.000 n=10+8)
ParseJSON/escape=40.00%/16384-32   73.3MB/s ± 1%  127.3MB/s ± 1%   +73.68%  (p=0.000 n=10+9)
ParseJSON/escape=40.00%/65536-32   73.4MB/s ± 1%  130.3MB/s ± 0%   +77.40%  (p=0.000 n=10+8)
ParseJSON/escape=40.00%/262144-32  74.6MB/s ± 0%  129.6MB/s ± 0%   +73.63%  (p=0.000 n=7+9)
ParseJSON/escape=40.00%/524288-32  75.6MB/s ± 1%  132.6MB/s ± 0%   +75.37%  (p=0.000 n=10+10)

name                               old alloc/op   new alloc/op    delta
ParseJSON/escape=0.00%/128-32        2.44kB ± 0%     1.06kB ± 0%   -56.42%  (p=0.000 n=10+10)
ParseJSON/escape=0.00%/512-32        3.25kB ± 0%     1.32kB ± 0%   -59.26%  (p=0.000 n=10+10)
ParseJSON/escape=0.00%/4096-32       13.0kB ± 0%      4.0kB ± 0%   -69.08%  (p=0.000 n=10+10)
ParseJSON/escape=0.00%/16384-32      45.3kB ± 0%     12.8kB ± 0%   -71.71%  (p=0.000 n=10+8)
ParseJSON/escape=0.00%/65536-32       177kB ± 0%       49kB ± 0%   -72.17%  (p=0.000 n=10+10)
ParseJSON/escape=0.00%/262144-32      717kB ± 0%      191kB ± 0%   -73.35%  (p=0.000 n=10+10)
ParseJSON/escape=0.00%/524288-32     1.40MB ± 0%     0.37MB ± 0%   -73.31%  (p=0.000 n=9+10)
ParseJSON/escape=3.00%/128-32        2.56kB ± 0%     1.16kB ± 0%   -54.83%  (p=0.000 n=10+10)
ParseJSON/escape=3.00%/512-32        3.74kB ± 0%     1.71kB ± 0%   -54.19%  (p=0.000 n=10+10)
ParseJSON/escape=3.00%/4096-32       16.7kB ± 0%      7.1kB ± 0%   -57.71%  (p=0.000 n=10+10)
ParseJSON/escape=3.00%/16384-32      59.3kB ± 0%     24.6kB ± 0%   -58.47%  (p=0.000 n=9+10)
ParseJSON/escape=3.00%/65536-32       236kB ± 0%       99kB ± 0%   -58.11%  (p=0.000 n=8+10)
ParseJSON/escape=3.00%/262144-32      940kB ± 0%      376kB ± 0%   -60.02%  (p=0.000 n=8+10)
ParseJSON/escape=3.00%/524288-32     1.83MB ± 0%     0.75MB ± 0%   -59.16%  (p=0.000 n=10+9)
ParseJSON/escape=40.00%/128-32       2.62kB ± 0%     1.16kB ± 0%   -55.62%  (p=0.000 n=10+10)
ParseJSON/escape=40.00%/512-32       4.17kB ± 0%     1.74kB ± 0%   -58.28%  (p=0.000 n=10+10)
ParseJSON/escape=40.00%/4096-32      19.7kB ± 0%      7.0kB ± 0%   -64.32%  (p=0.000 n=10+10)
ParseJSON/escape=40.00%/16384-32     72.7kB ± 0%     24.9kB ± 0%   -65.80%  (p=0.000 n=10+10)
ParseJSON/escape=40.00%/65536-32      293kB ± 0%      101kB ± 0%   -65.55%  (p=0.000 n=9+9)
ParseJSON/escape=40.00%/262144-32    1.12MB ± 0%     0.37MB ± 1%   -66.76%  (p=0.000 n=9+10)
ParseJSON/escape=40.00%/524288-32    2.42MB ± 1%     0.74MB ± 0%   -69.48%  (p=0.000 n=10+10)

name                               old allocs/op  new allocs/op   delta
ParseJSON/escape=0.00%/128-32          39.0 ± 0%       29.0 ± 0%   -25.64%  (p=0.000 n=10+10)
ParseJSON/escape=0.00%/512-32          39.0 ± 0%       29.0 ± 0%   -25.64%  (p=0.000 n=10+10)
ParseJSON/escape=0.00%/4096-32         41.0 ± 0%       29.0 ± 0%   -29.27%  (p=0.000 n=10+10)
ParseJSON/escape=0.00%/16384-32        41.0 ± 0%       29.0 ± 0%   -29.27%  (p=0.000 n=10+10)
ParseJSON/escape=0.00%/65536-32        43.0 ± 0%       29.0 ± 0%   -32.56%  (p=0.000 n=10+10)
ParseJSON/escape=0.00%/262144-32       45.0 ± 0%       29.0 ± 0%   -35.56%  (p=0.000 n=10+10)
ParseJSON/escape=0.00%/524288-32       44.0 ± 0%       29.0 ± 0%   -34.09%  (p=0.002 n=8+10)
ParseJSON/escape=3.00%/128-32          40.0 ± 0%       30.0 ± 0%   -25.00%  (p=0.000 n=10+10)
ParseJSON/escape=3.00%/512-32          41.0 ± 0%       30.0 ± 0%   -26.83%  (p=0.000 n=10+10)
ParseJSON/escape=3.00%/4096-32         42.0 ± 0%       30.0 ± 0%   -28.57%  (p=0.000 n=10+10)
ParseJSON/escape=3.00%/16384-32        43.0 ± 0%       30.0 ± 0%   -30.23%  (p=0.000 n=10+10)
ParseJSON/escape=3.00%/65536-32        44.0 ± 0%       30.0 ± 0%   -31.82%  (p=0.000 n=10+10)
ParseJSON/escape=3.00%/262144-32       46.0 ± 0%       30.0 ± 0%   -34.78%  (p=0.000 n=10+10)
ParseJSON/escape=3.00%/524288-32       45.7 ± 2%       30.0 ± 0%   -34.35%  (p=0.000 n=10+10)
ParseJSON/escape=40.00%/128-32         40.0 ± 0%       30.0 ± 0%   -25.00%  (p=0.000 n=10+10)
ParseJSON/escape=40.00%/512-32         41.0 ± 0%       31.0 ± 0%   -24.39%  (p=0.000 n=10+10)
ParseJSON/escape=40.00%/4096-32        42.0 ± 0%       30.0 ± 0%   -28.57%  (p=0.000 n=10+10)
ParseJSON/escape=40.00%/16384-32       43.0 ± 0%       30.0 ± 0%   -30.23%  (p=0.000 n=10+10)
ParseJSON/escape=40.00%/65536-32       45.0 ± 0%       30.0 ± 0%   -33.33%  (p=0.000 n=10+10)
ParseJSON/escape=40.00%/262144-32      45.0 ± 0%       30.0 ± 0%   -33.33%  (p=0.000 n=10+10)
ParseJSON/escape=40.00%/524288-32      46.0 ± 0%       30.0 ± 0%   -34.78%  (p=0.000 n=9+10)
```

miretskiy commented:
It should be noted that a previous attempt using json-iter (https://github.com/cockroachdb/cockroach/pull/89117)
was also pretty good.  This implementation, however is still faster (at least 10%) and has much better allocations
behavior.

miretskiy commented:
A bit more changes here to avoid initial input copy -- speed up relative to previous version:
```
ParseJSON/escape=0.00%/128-32        2.68µs ± 2%    2.53µs ± 1%   -5.65%  (p=0.000 n=9+9)
ParseJSON/escape=0.00%/512-32        3.02µs ± 2%    2.83µs ± 1%   -6.44%  (p=0.000 n=10+9)
ParseJSON/escape=0.00%/4096-32       6.92µs ± 3%    6.31µs ± 1%   -8.84%  (p=0.000 n=10+9)
ParseJSON/escape=0.00%/16384-32      18.0µs ± 1%    16.5µs ± 1%   -8.33%  (p=0.000 n=9+9)
ParseJSON/escape=0.00%/65536-32      61.9µs ± 1%    52.9µs ± 8%  -14.62%  (p=0.000 n=9+10)
ParseJSON/escape=0.00%/262144-32      230µs ± 2%     197µs ± 8%  -14.28%  (p=0.000 n=9+10)
ParseJSON/escape=0.00%/524288-32      427µs ± 3%     359µs ± 3%  -15.95%  (p=0.000 n=10+9)

ParseJSON/escape=0.00%/128-32      62.7MB/s ± 2%  66.4MB/s ± 1%   +5.97%  (p=0.000 n=9+9)
ParseJSON/escape=0.00%/512-32       141MB/s ± 2%   151MB/s ± 1%   +6.87%  (p=0.000 n=10+9)
ParseJSON/escape=0.00%/4096-32      422MB/s ± 3%   463MB/s ± 1%   +9.67%  (p=0.000 n=10+9)
ParseJSON/escape=0.00%/16384-32     628MB/s ± 1%   685MB/s ± 1%   +9.08%  (p=0.000 n=9+9)
ParseJSON/escape=0.00%/65536-32     725MB/s ± 1%   851MB/s ± 9%  +17.38%  (p=0.000 n=9+10)
ParseJSON/escape=0.00%/262144-32    808MB/s ± 1%   945MB/s ± 8%  +16.97%  (p=0.000 n=9+10)
ParseJSON/escape=0.00%/524288-32    863MB/s ± 3%  1027MB/s ± 3%  +19.03%  (p=0.000 n=10+9)
```

Escaped string cases show modest improvement when few escape characters present (~4%), dropping
to basically a wash with large inputs.  Nonetheless, this latest changed removed some of the code, so a net
win.

miretskiy commented:
Fixes #89998

miretskiy commented:
Most recent stats against current implementation as as follows:
```
name                               old time/op    new time/op     delta
ParseJSON/escape=0.00%/128-32        5.50µs ± 5%     2.51µs ± 3%   -54.32%  (p=0.000 n=9+8)
ParseJSON/escape=0.00%/512-32        7.31µs ± 6%     2.77µs ± 1%   -62.04%  (p=0.000 n=10+8)
ParseJSON/escape=0.00%/4096-32       26.5µs ± 1%      5.8µs ± 2%   -77.95%  (p=0.000 n=8+8)
ParseJSON/escape=0.00%/16384-32      85.1µs ± 1%     14.7µs ± 2%   -82.72%  (p=0.000 n=9+9)
ParseJSON/escape=0.00%/262144-32     1.35ms ± 1%     0.18ms ± 2%   -86.66%  (p=0.000 n=8+10)
ParseJSON/escape=0.00%/524288-32     2.62ms ± 1%     0.33ms ± 1%   -87.53%  (p=0.000 n=9+8)
ParseJSON/escape=3.00%/128-32        5.83µs ± 3%     2.91µs ± 1%   -50.04%  (p=0.000 n=9+9)
ParseJSON/escape=3.00%/512-32        8.63µs ± 0%     4.38µs ± 1%   -49.19%  (p=0.000 n=7+9)
ParseJSON/escape=3.00%/4096-32       34.0µs ± 1%     18.5µs ± 1%   -45.69%  (p=0.000 n=9+10)
ParseJSON/escape=3.00%/16384-32       110µs ± 2%       65µs ± 1%   -40.57%  (p=0.000 n=9+9)
ParseJSON/escape=3.00%/262144-32     1.76ms ± 7%     1.01ms ± 1%   -42.67%  (p=0.000 n=8+9)
ParseJSON/escape=3.00%/524288-32     3.44ms ± 5%     2.00ms ± 1%   -41.89%  (p=0.000 n=10+8)
ParseJSON/escape=40.00%/128-32       6.52µs ± 3%     3.29µs ± 0%   -49.61%  (p=0.000 n=10+9)
ParseJSON/escape=40.00%/512-32       11.7µs ± 1%      5.2µs ± 1%   -55.45%  (p=0.000 n=8+8)
ParseJSON/escape=40.00%/4096-32      55.6µs ± 1%     25.7µs ± 0%   -53.80%  (p=0.000 n=10+8)
ParseJSON/escape=40.00%/16384-32      204µs ± 1%       91µs ± 0%   -55.26%  (p=0.000 n=10+8)
ParseJSON/escape=40.00%/262144-32    3.18ms ± 0%     1.44ms ± 1%   -54.70%  (p=0.000 n=7+8)
ParseJSON/escape=40.00%/524288-32    6.77ms ± 1%     2.89ms ± 3%   -57.24%  (p=0.000 n=10+10)

name                               old speed      new speed       delta
ParseJSON/escape=0.00%/128-32      30.6MB/s ± 4%   66.9MB/s ± 3%  +118.89%  (p=0.000 n=9+8)
ParseJSON/escape=0.00%/512-32      58.4MB/s ± 6%  153.6MB/s ± 1%  +163.12%  (p=0.000 n=10+8)
ParseJSON/escape=0.00%/4096-32      110MB/s ± 1%    501MB/s ± 2%  +353.49%  (p=0.000 n=8+8)
ParseJSON/escape=0.00%/16384-32     133MB/s ± 1%    769MB/s ± 2%  +478.79%  (p=0.000 n=9+9)
ParseJSON/escape=0.00%/262144-32    138MB/s ± 1%   1038MB/s ± 2%  +654.61%  (p=0.000 n=8+10)
ParseJSON/escape=0.00%/524288-32    139MB/s ± 1%   1110MB/s ± 0%  +699.16%  (p=0.000 n=9+8)
ParseJSON/escape=3.00%/128-32      29.5MB/s ± 3%   59.4MB/s ± 1%  +101.30%  (p=0.000 n=9+9)
ParseJSON/escape=3.00%/512-32      52.4MB/s ± 0%  103.3MB/s ± 1%   +97.26%  (p=0.000 n=7+9)
ParseJSON/escape=3.00%/4096-32     90.0MB/s ± 1%  165.6MB/s ± 1%   +83.94%  (p=0.000 n=9+10)
ParseJSON/escape=3.00%/16384-32     107MB/s ± 2%    183MB/s ± 1%   +70.63%  (p=0.000 n=9+9)
ParseJSON/escape=3.00%/262144-32    108MB/s ± 5%    190MB/s ± 1%   +75.70%  (p=0.000 n=7+9)
ParseJSON/escape=3.00%/524288-32    111MB/s ± 5%    190MB/s ± 1%   +71.68%  (p=0.000 n=10+9)
ParseJSON/escape=40.00%/128-32     29.9MB/s ± 3%   61.7MB/s ± 1%  +106.44%  (p=0.000 n=10+10)
ParseJSON/escape=40.00%/512-32     47.4MB/s ± 1%  100.1MB/s ± 1%  +111.12%  (p=0.000 n=8+8)
ParseJSON/escape=40.00%/4096-32    68.5MB/s ± 1%  152.9MB/s ± 0%  +123.10%  (p=0.000 n=10+8)
ParseJSON/escape=40.00%/16384-32   73.3MB/s ± 1%  164.0MB/s ± 0%  +123.80%  (p=0.000 n=10+8)
ParseJSON/escape=40.00%/262144-32  74.6MB/s ± 0%  168.7MB/s ± 0%  +126.05%  (p=0.000 n=7+8)
ParseJSON/escape=40.00%/524288-32  75.6MB/s ± 1%  168.1MB/s ± 1%  +122.36%  (p=0.000 n=10+9)

name                               old alloc/op   new alloc/op    delta
ParseJSON/escape=0.00%/128-32        2.44kB ± 0%     0.97kB ± 0%   -60.27%  (p=0.000 n=10+10)
ParseJSON/escape=0.00%/512-32        3.25kB ± 0%     1.23kB ± 0%   -62.13%  (p=0.000 n=10+10)
ParseJSON/escape=0.00%/4096-32       13.0kB ± 0%      3.9kB ± 0%   -69.82%  (p=0.000 n=10+10)
ParseJSON/escape=0.00%/16384-32      45.3kB ± 0%     12.7kB ± 0%   -71.92%  (p=0.000 n=10+10)
ParseJSON/escape=0.00%/262144-32      717kB ± 0%      192kB ± 0%   -73.26%  (p=0.000 n=10+9)
ParseJSON/escape=0.00%/524288-32     1.40MB ± 0%     0.37MB ± 0%   -73.76%  (p=0.000 n=9+9)
ParseJSON/escape=3.00%/128-32        2.56kB ± 0%     0.97kB ± 0%   -62.14%  (p=0.000 n=10+10)
ParseJSON/escape=3.00%/512-32        3.74kB ± 0%     1.25kB ± 0%   -66.70%  (p=0.000 n=10+10)
ParseJSON/escape=3.00%/4096-32       16.7kB ± 0%      3.9kB ± 0%   -76.55%  (p=0.000 n=10+10)
ParseJSON/escape=3.00%/16384-32      59.3kB ± 0%     12.9kB ± 0%   -78.27%  (p=0.000 n=9+9)
ParseJSON/escape=3.00%/262144-32      940kB ± 0%      190kB ± 0%   -79.83%  (p=0.001 n=8+6)
ParseJSON/escape=3.00%/524288-32     1.83MB ± 0%     0.37MB ± 0%   -79.84%  (p=0.000 n=10+8)
ParseJSON/escape=40.00%/128-32       2.62kB ± 0%     1.00kB ± 0%   -62.03%  (p=0.000 n=10+10)
ParseJSON/escape=40.00%/512-32       4.17kB ± 0%     1.22kB ± 0%   -70.80%  (p=0.000 n=10+10)
ParseJSON/escape=40.00%/4096-32      19.7kB ± 0%      4.0kB ± 0%   -79.66%  (p=0.000 n=10+10)
ParseJSON/escape=40.00%/16384-32     72.7kB ± 0%     12.9kB ± 0%   -82.32%  (p=0.000 n=10+8)
ParseJSON/escape=40.00%/262144-32    1.12MB ± 0%     0.19MB ± 1%   -82.98%  (p=0.000 n=9+10)
ParseJSON/escape=40.00%/524288-32    2.42MB ± 1%     0.38MB ± 0%   -84.51%  (p=0.000 n=10+9)

name                               old allocs/op  new allocs/op   delta
ParseJSON/escape=0.00%/128-32          39.0 ± 0%       26.0 ± 0%   -33.33%  (p=0.000 n=10+10)
ParseJSON/escape=0.00%/512-32          39.0 ± 0%       26.0 ± 0%   -33.33%  (p=0.000 n=10+10)
ParseJSON/escape=0.00%/4096-32         41.0 ± 0%       26.0 ± 0%   -36.59%  (p=0.000 n=10+10)
ParseJSON/escape=0.00%/16384-32        41.0 ± 0%       26.0 ± 0%   -36.59%  (p=0.000 n=10+10)
ParseJSON/escape=0.00%/262144-32       45.0 ± 0%       26.0 ± 0%   -42.22%  (p=0.000 n=10+10)
ParseJSON/escape=0.00%/524288-32       44.0 ± 0%       26.0 ± 0%   -40.91%  (p=0.002 n=8+10)
ParseJSON/escape=3.00%/128-32          40.0 ± 0%       26.0 ± 0%   -35.00%  (p=0.000 n=10+10)
ParseJSON/escape=3.00%/512-32          41.0 ± 0%       26.0 ± 0%   -36.59%  (p=0.000 n=10+10)
ParseJSON/escape=3.00%/4096-32         42.0 ± 0%       26.0 ± 0%   -38.10%  (p=0.000 n=10+10)
ParseJSON/escape=3.00%/16384-32        43.0 ± 0%       26.0 ± 0%   -39.53%  (p=0.000 n=10+10)
ParseJSON/escape=3.00%/262144-32       46.0 ± 0%       26.0 ± 0%   -43.48%  (p=0.000 n=10+10)
ParseJSON/escape=3.00%/524288-32       45.7 ± 2%       26.0 ± 0%   -43.11%  (p=0.000 n=10+10)
ParseJSON/escape=40.00%/128-32         40.0 ± 0%       27.0 ± 0%   -32.50%  (p=0.000 n=10+10)
ParseJSON/escape=40.00%/512-32         41.0 ± 0%       26.0 ± 0%   -36.59%  (p=0.000 n=10+10)
ParseJSON/escape=40.00%/4096-32        42.0 ± 0%       26.0 ± 0%   -38.10%  (p=0.000 n=10+10)
ParseJSON/escape=40.00%/16384-32       43.0 ± 0%       26.0 ± 0%   -39.53%  (p=0.000 n=10+10)
ParseJSON/escape=40.00%/262144-32      45.0 ± 0%       26.3 ± 3%   -41.56%  (p=0.000 n=10+10)
ParseJSON/escape=40.00%/524288-32      46.0 ± 0%       26.0 ± 0%   -43.48%  (p=0.000 n=9+10)
```

HonoreDB reviewed (approved):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/89884)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @jordanlewis, @mgartner, @miretskiy, and @yuzefovich)

___
*[`pkg/util/json/tokenizer/scanner.go` line 118 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/89884#-NEr2pWJ-sxAo9M8spPX:-NEr2pWK0z6PUYlU-yfa:b9jdkfk) ([raw file](https://github.com/cockroachdb/cockroach/blob/b34b013c6541eec7865ad6d52cd2617437bcfd6e/pkg/util/json/tokenizer/scanner.go#L118)):*
> ```Go
> 		switch c {
> 		case True:
> 			return s.next(validateToken(s.buf(), "true"))
> ```

Here we're converting a byte slice to a string, comparing the two, then passing the length of the slice to next(), which is guaranteed to return the byte slice again. It feels like there should be a way to optimize this even further? 
___
*[`pkg/util/json/tokenizer/scanner.go` line 180 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/89884#-NEr4aETEZzUY37LaHaq:-NEr4aETEZzUY37LaHar:b73s60e) ([raw file](https://github.com/cockroachdb/cockroach/blob/b34b013c6541eec7865ad6d52cd2617437bcfd6e/pkg/util/json/tokenizer/scanner.go#L180)):*
> ```Go
> 
> 	// Fast path: string does not have escape sequences.
> 	for _, c := range w {
> ```

Matt had an interesting idea here to optimize this using bit twiddling (probably via bytes.ContainsAny) that might be worth trying at some point. e.g. check if the next 8 bytes contain any backslashes, quotes, or control characters, and if not we don't need to range over them at all.
___
*[`pkg/util/json/tokenizer/scanner.go` line 201 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/89884#-NEr8geQ0dfE3GAQ4dMY:-NEr8geQ0dfE3GAQ4dMZ:b23546z) ([raw file](https://github.com/cockroachdb/cockroach/blob/b34b013c6541eec7865ad6d52cd2617437bcfd6e/pkg/util/json/tokenizer/scanner.go#L201)):*
> ```Go
> }
> 
> var bufferPool = sync.Pool{New: func() interface{} { return &buffer{} }}
> ```

This is probably a separate project but we might benefit from a pool+grow implementation that let grow use the pool rather than always allocating new memory.
___
*[`pkg/util/json/tokenizer/scanner.go` line 321 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/89884#-NErD5mqEM5CdRH3eA0Z:-NErD5mqEM5CdRH3eA0_:b-z8nfqs) ([raw file](https://github.com/cockroachdb/cockroach/blob/b34b013c6541eec7865ad6d52cd2617437bcfd6e/pkg/util/json/tokenizer/scanner.go#L321)):*
> ```Go
> 	var state uint8 = begin
> 
> 	// handle the case that the first character is a hyphen
> ```

why isn't this just part of case begin?
___
*[`pkg/util/json/tokenizer/scanner.go` line 429 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/89884#-NErF-N02L2FYEnKjM1k:-NErF-N02L2FYEnKjM1l:b-jdzhly) ([raw file](https://github.com/cockroachdb/cockroach/blob/b34b013c6541eec7865ad6d52cd2617437bcfd6e/pkg/util/json/tokenizer/scanner.go#L429)):*
> ```Go
> 		return -1
> 	}
> 	return r1*(1<<12) + r2*(1<<8) + r3*(1<<4) + r4
> ```

`(r1 << 12) | (r2 << 8) | (r3 << 4) | r4)`?

Or for bonus micro-optimization, just have four hexTables with the values pre-shifted the correct amounts and do `hexTable1[r1] | hexTable2[r2] | hexTable3[r3] | hexTable4[4]`.


<!-- Sent from Reviewable.io -->

miretskiy commented:
> Here we're converting a byte slice to a string, comparing the two, then passing the length of the slice to next(), which is guaranteed to return the byte slice again. It feels like there should be a way to optimize this even further?

I don't think it's problematic or needs to be optimized. This is a constant string, and iterating over it is
fine -- we are comparing bytes to bytes, and returning bytes.

> Matt had an interesting idea here to optimize this using bit twiddling (probably via bytes.ContainsAny) that might be worth trying at some point. e.g. check if the next 8 bytes contain any backslashes, quotes, or control characters, and if not we don't need to range over them at all.

Might be worth experimenting with; I have lots of doubts though.  The problem is that I need to search for more than \\; I need to search for any unescaped control characters as well.  Since I will have a set of 33 characters to search for (all control chars, plus \\, and maybe even 34 characters to include ") -- then contains any implementation will fall back to
```
	if as, isASCII := makeASCIISet(chars); isASCII {
			for i, c := range s {
				if as.contains(c) {
					return i
				}
			}
			return -1
		}
```
Since i will be invoking makeASCIISet for each string, suspect this implementation will be substantially slower.
I can, of course build the set myself -- but then you see the problem -- i'm just iterating and checking if something is in the set... But, i'm not sure that comparison for e.g. `c < ' '` are that much slower.  At any rate, might be worth checking, but as I said, I have my doubts.

> why isn't this just part of case begin?
Not sure -- this is the part of the code I borrowed, but it appears that if you made it part of begin, you'd need to intro
another case because you want to catch error cases, such as "--".

> (r1 << 12) | (r2 << 8) | (r3 << 4) | r4)?
> Or for bonus micro-optimization, just have four hexTables with the values pre-shifted the correct amounts and do hexTable1[r1] | hexTable2[r2] | hexTable3[r3] | hexTable4[4].

might be worth it.. Or have `var hexTable [256][4]rune`

mgartner reviewed (commented):
Very cool! I did a first pass, and overall it looks great to me. I'd like to take another pass early next week, but left some comments. Also have some high level comments:

* The usage of the term "lexer" as in `useJSONLexer` and elsewhere is confusing to me - it's doing more than lexing, it's parsing. Is there a better name we can use?
* In the interest of safety, I think it'd be nice to ship this with a session setting to revert to the standard library parser. Given that it's in the std lib, it won't be painful keeping the slower alternative until 23.2 or beyond when we have confidence that the fast parser doesn't break production workloads. I don't think this setting needs to be added in this PR, though.

> Reviewed 13 of 17 files at r1.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/89884#-:-NEvzH-54r2BOhzmP9SR:bq9iptv)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @jordanlewis, @miretskiy, and @yuzefovich)

___
*[`pkg/util/json/config.go` line 48 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/89884#-NEvqbHE3E2P1vcE1Zu7:-NEvqbHE3E2P1vcE1Zu8:bmzmx56) ([raw file](https://github.com/cockroachdb/cockroach/blob/b34b013c6541eec7865ad6d52cd2617437bcfd6e/pkg/util/json/config.go#L48)):*
> ```Go
> // parseJSONImplType is the implementation library
> // used to parse JSON.
> type parseJSONImplType int
> ```

nit: `int8` would make `parseConfig` fit in a single word.
___
*[`pkg/util/json/tokenizer/README.md` line 1 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/89884#-NEvxHID3WWGMeJSUH8c:-NEvxHID3WWGMeJSUH8d:b-lv0vmi) ([raw file](https://github.com/cockroachdb/cockroach/blob/b34b013c6541eec7865ad6d52cd2617437bcfd6e/pkg/util/json/tokenizer/README.md#L1)):*
> ```Markdown
> # pkg/json [![GoDoc](https://godoc.org/github.com/pkg/json?status.svg)](https://godoc.org/github.com/pkg/json)
> ```

I don't think we need this copy of the README from `pkg/json`, do we?
___
*[`pkg/util/json/tokenizer/README_crdb.md` line 4 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/89884#-NEvx1e05aGBptQOmhVZ:-NEvx1e05aGBptQOmhV_:b-q0eaiq) ([raw file](https://github.com/cockroachdb/cockroach/blob/b34b013c6541eec7865ad6d52cd2617437bcfd6e/pkg/util/json/tokenizer/README_crdb.md#L4)):*
> ```Markdown
> 
> These files are a fork of `pkg/json` package.
> - `scanner.go`, `decoder.go`, and `reader.go` were copied from pkg/json
> ```

Might be nice to include the un-altered versions of these files in the first commit, and then the altered in a second commit so we can see the changes you made in detail.
___
*[`pkg/sql/sem/tree/json_test.go` line 120 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/89884#-NEvwi9i6pYhFr8A30lE:-NEvwi9i6pYhFr8A30lF:b8fiy9t) ([raw file](https://github.com/cockroachdb/cockroach/blob/b34b013c6541eec7865ad6d52cd2617437bcfd6e/pkg/sql/sem/tree/json_test.go#L120)):*
> ```Go
> 	b.ResetTimer()
> 
> 	//parseOpts := []jsonb.ParseOption{jsonb.WithGoStandardParser()}
> ```

nit: looks like leftover?
___
*[`pkg/util/json/json_test.go` line 312 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/89884#-NEvvjrO-AD4EvESxgMp:-NEvvjrO-AD4EvESxgMq:bfuah23) ([raw file](https://github.com/cockroachdb/cockroach/blob/b34b013c6541eec7865ad6d52cd2617437bcfd6e/pkg/util/json/json_test.go#L312)):*
> ```Go
> 	testCases := []testCaseDef{
> 		testCase(`true false`, trailingChars, useStdGoJSON),
> 		testCase(`true false`, trailingChars, useJSONLexer),
> ```

We could make `testCaseDef` a `struct {input string, ..., expectStdGoJSONErr, expectJSONLexerErr string}` so it's impossible to forget to test one of the variants.
___
*[`pkg/util/json/json_test.go` line 380 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/89884#-NEvvYeOAcbJc4_SAXnv:-NEvvYeOAcbJc4_SAXnw:bu7kmpm) ([raw file](https://github.com/cockroachdb/cockroach/blob/b34b013c6541eec7865ad6d52cd2617437bcfd6e/pkg/util/json/json_test.go#L380)):*
> ```Go
> 			//if !strings.Contains(expectErr.Error(), tc.msg[impl]) {
> 			//	t.Fatalf("expected error message to be '%s', but was '%s'", tc.msg[impl], expectErr.Error())
> 			//}
> ```

Did you intend to delete these lines?


<!-- Sent from Reviewable.io -->

Commit:
sql: Fork a pkg/json package

Add a fork of pkg/json package (https://github.com/pkg/json).
This is just an initial checkin so that we can modify
and adopt this package for internal use.

The changes made include removal of some of the functionality
that will not be needed, fix ups around internal conventions/lint, etc.

Release note: None

miretskiy commented:
> * The usage of the term "lexer" as in `useJSONLexer` and elsewhere is confusing to me - it's doing more than lexing, it's parsing. Is there a better name we can use?

renamed to `fastJSONParser`. 

> * In the interest of safety, I think it'd be nice to ship this with a session setting to revert to the standard library parser. Given that it's in the std lib, it won't be painful keeping the slower alternative until 23.2 or beyond when we have confidence that the fast parser doesn't break production workloads. I don't think this setting needs to be added in this PR, though.

As discussed, doing so is not trivial.  We are pushing this code out early in the release cycle -- so, we'll have plenty
of time to bake.

> Might be nice to include the un-altered versions of these files in the first commit, and then the altered in a second commit so we can see the changes you made in detail.

Done -- added 1st commit containing almost unmodified pkg/json package (also included some of the tests).

> Did you intend to delete these lines?
Yup -- forgot to delete; I split up the test so that different implementation have different
test cases since new implementation has *much* nicer error messages.

> nit: looks like leftover?
Yup.

miretskiy commented:
> We could make testCaseDef a struct {input string, ..., expectStdGoJSONErr, expectJSONLexerErr string} so it's impossible to forget to test one of the variants.

I had it at some point, and I just found things to be even harder to read; also.. maybe we'll get rid of the other parser at some point.  Do you feel strongly?

HonoreDB commented:
>The problem is that I need to search for more than \; I need to search for any unescaped control characters as well.
In theory you can do this part pretty quickly too because you're just looking for bytes that don't have any bits in `255^31` set. I wrote [this](https://go.dev/play/p/tbmQkJlUNPk) over the weekend to demonstrate, don't know if it's portable or worth it.

miretskiy commented:
> In theory you can do this part pretty quickly too because you're just looking for bytes that don't have any bits in `255^31` set. I wrote [this](https://go.dev/play/p/tbmQkJlUNPk) over the weekend to demonstrate, don't know if it's portable or worth it.

Feel free to benchmark -- certainly, LOTS of complex code (over a simple loop) should have significantly better performance to be worth it.

knz reviewed (commented):
> Reviewed 3 of 21 files at r3.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/89884)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @jordanlewis, @mgartner, @miretskiy, and @yuzefovich)

___
*[`pkg/util/json/tokenizer/scanner.go` line 103 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/89884#-NF9_HHP4gW_maa-xAbd:-NF9_HHP4gW_maa-xAbe:b-1o9c3i) ([raw file](https://github.com/cockroachdb/cockroach/blob/38053b635c8ef3f2acf9db508a9e6ac08b240d9c/pkg/util/json/tokenizer/scanner.go#L103)):*
> ```Go
> 	w := s.buf()
> 	for pos, c := range w {
> 		// strip any leading whitespace.
> ```

nit: capital at end of sentence.
___
*[`pkg/util/json/tokenizer/scanner.go` line 108 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/89884#-NF9_JVr9a6FX1q3q_zP:-NF9_JVr9a6FX1q3q_zQ:b1i56q3) ([raw file](https://github.com/cockroachdb/cockroach/blob/38053b635c8ef3f2acf9db508a9e6ac08b240d9c/pkg/util/json/tokenizer/scanner.go#L108)):*
> ```Go
> 		}
> 
> 		// simple case
> ```

nit: capital, period.
___
*[`pkg/util/json/tokenizer/scanner.go` line 126 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/89884#-NF9_LBr8Ha0xHd4IdGC:-NF9_LBr8Ha0xHd4IdGD:b-7ropln) ([raw file](https://github.com/cockroachdb/cockroach/blob/38053b635c8ef3f2acf9db508a9e6ac08b240d9c/pkg/util/json/tokenizer/scanner.go#L126)):*
> ```Go
> 			return s.parseString()
> 		default:
> 			// ensure the number is correct.
> ```

ditto
___
*[`pkg/util/json/tokenizer/scanner.go` line 131 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/89884#-NF9_M-d1CrNXYcYQXUf:-NF9_M-d1CrNXYcYQXUg:b-7ropln) ([raw file](https://github.com/cockroachdb/cockroach/blob/38053b635c8ef3f2acf9db508a9e6ac08b240d9c/pkg/util/json/tokenizer/scanner.go#L131)):*
> ```Go
> 	}
> 
> 	// it's all whitespace, ignore it
> ```

ditto
___
*[`pkg/util/json/tokenizer/scanner.go` line 194 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/89884#-NF9a51GBZZ753ibFhuZ:-NF9a51GBZZ753ibFhu_:b-83ohoi) ([raw file](https://github.com/cockroachdb/cockroach/blob/38053b635c8ef3f2acf9db508a9e6ac08b240d9c/pkg/util/json/tokenizer/scanner.go#L194)):*
> ```Go
> 
> 		if c < ' ' {
> 			// Unescaped controlled characters not allowed.
> ```

0x1f (ascii 127) is also a control character.
___
*[`pkg/util/json/tokenizer/scanner.go` line 230 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/89884#-NF9a9Wq8pNRx7N3Ht5C:-NF9a9Wq8pNRx7N3Ht5D:b-83ohoi) ([raw file](https://github.com/cockroachdb/cockroach/blob/38053b635c8ef3f2acf9db508a9e6ac08b240d9c/pkg/util/json/tokenizer/scanner.go#L230)):*
> ```Go
> 			wp++
> 		case c < ' ':
> 			// Control characters must be escaped.
> ```

0x1f (ascii 127) is also a control character.
___
*[`pkg/util/json/tokenizer/scanner.go` line 318 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/89884#-NF9b6W1-l2cWB9YyNgV:-NF9b6W1-l2cWB9YyNgW:b1i56q3) ([raw file](https://github.com/cockroachdb/cockroach/blob/38053b635c8ef3f2acf9db508a9e6ac08b240d9c/pkg/util/json/tokenizer/scanner.go#L318)):*
> ```Go
> 	pos := 0
> 	w := s.buf()
> 	// int vs uint8 costs 10% on canada.json
> ```

nit: capital, period.
___
*[`pkg/util/json/tokenizer/scanner.go` line 321 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/89884#-NF9b94T2GiI-BrzCLah:-NF9b94T2GiI-BrzCLai:b1i56q3) ([raw file](https://github.com/cockroachdb/cockroach/blob/38053b635c8ef3f2acf9db508a9e6ac08b240d9c/pkg/util/json/tokenizer/scanner.go#L321)):*
> ```Go
> 	var state uint8 = begin
> 
> 	// handle the case that the first character is a hyphen
> ```

nit: capital, period.
___
*[`pkg/util/json/tokenizer/scanner.go` line 335 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/89884#-NF9bAQkANRbXwQGOvL3:-NF9bAQkANRbXwQGOvL4:b1i56q3) ([raw file](https://github.com/cockroachdb/cockroach/blob/38053b635c8ef3f2acf9db508a9e6ac08b240d9c/pkg/util/json/tokenizer/scanner.go#L335)):*
> ```Go
> 				state = leadingzero
> 			} else {
> 				// error
> ```

nit: capital, period.
___
*[`pkg/util/json/tokenizer/scanner.go` line 340 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/89884#-NF9bB1H8flWyCUrdadM:-NF9bB1H8flWyCUrdadN:b1i56q3) ([raw file](https://github.com/cockroachdb/cockroach/blob/38053b635c8ef3f2acf9db508a9e6ac08b240d9c/pkg/util/json/tokenizer/scanner.go#L340)):*
> ```Go
> 		case anydigit1:
> 			if elem >= '0' && elem <= '9' {
> 				// stay in this state
> ```

nit: capital, period.
___
*[`pkg/util/json/tokenizer/scanner.go` line 358 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/89884#-NF9bBqp7s-ddl85VXho:-NF9bBqp7s-ddl85VXhp:b-7ropln) ([raw file](https://github.com/cockroachdb/cockroach/blob/38053b635c8ef3f2acf9db508a9e6ac08b240d9c/pkg/util/json/tokenizer/scanner.go#L358)):*
> ```Go
> 				state = anydigit2
> 			} else {
> 				// error
> ```

ditto
___
*[`pkg/util/json/tokenizer/scanner.go` line 381 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/89884#-NF9bCbn0HyO1dL4X__M:-NF9bCbn0HyO1dL4X__N:b-7ropln) ([raw file](https://github.com/cockroachdb/cockroach/blob/38053b635c8ef3f2acf9db508a9e6ac08b240d9c/pkg/util/json/tokenizer/scanner.go#L381)):*
> ```Go
> 				break
> 			}
> 			// error
> ```

ditto
___
*[`pkg/util/json/tokenizer/scanner.go` line 391 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/89884#-NF9bDtP4dfb2nTmmzLF:-NF9bDtP4dfb2nTmmzLG:b-tvy9sw) ([raw file](https://github.com/cockroachdb/cockroach/blob/38053b635c8ef3f2acf9db508a9e6ac08b240d9c/pkg/util/json/tokenizer/scanner.go#L391)):*
> ```Go
> 	}
> 
> 	// end of the item. However, not necessarily an error. Make
> ```

nit: capital.
___
*[`pkg/util/json/tokenizer/scanner.go` line 397 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/89884#-NF9bF9c6cCMMUz4L1oE:-NF9bF9c6cCMMUz4L1oF:b-7ropln) ([raw file](https://github.com/cockroachdb/cockroach/blob/38053b635c8ef3f2acf9db508a9e6ac08b240d9c/pkg/util/json/tokenizer/scanner.go#L397)):*
> ```Go
> 		return pos
> 	default:
> 		// error otherwise, the number isn't complete.
> ```

ditto
___
*[`pkg/util/json/json_test.go` line 161 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/89884#-NF9bK0W3CduvIXOGKDM:-NF9bK0W3CduvIXOGKDN:bj7bbtl) ([raw file](https://github.com/cockroachdb/cockroach/blob/38053b635c8ef3f2acf9db508a9e6ac08b240d9c/pkg/util/json/json_test.go#L161)):*
> ```Go
> }
> 
> func TestJSONRoundTrip(t *testing.T) {
> ```

Please add some tests with ascii \x7f as well.
___
*[`pkg/util/json/json_test.go` line 293 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/89884#-NF9bOkK2NQvUfnt4JTp:-NF9bOkK2NQvUfnt4JTq:b-7ropln) ([raw file](https://github.com/cockroachdb/cockroach/blob/38053b635c8ef3f2acf9db508a9e6ac08b240d9c/pkg/util/json/json_test.go#L293)):*
> ```Go
> }
> 
> func TestJSONErrors(t *testing.T) {
> ```

ditto


<!-- Sent from Reviewable.io -->

miretskiy commented:
> nit: capital at end of sentence.
I could address all of these nits -- but these are from the forked code;
I tried to address only the things that make our linter scream... Do you feel strongly about those?

knz commented:
>  I could address all of these nits -- but these are from the forked code;
>  I tried to address only the things that make our linter scream... Do you feel strongly about those?

Yes. If you copy-paste code into our repo, the contract is that we are *taking over* its future maintainance, so it has to follow our own engineering guidelines. We did the same when we imported the other things.

If you don't want to take it over, including all our engineering requirements, then you can vendor the code instead.

miretskiy commented:
> 0x1f (ascii 127) is also a control character.

Assume you meant 0x7f -- anyway, verified that go encoding/json does not require this character
to be escaped; so, keeping same behavior as go standard library and also, json rfc: 
```
All Unicode characters may be placed within the
   quotation marks, except for the characters that MUST be escaped:
   quotation mark, reverse solidus, and the control characters (U+0000
   through U+001F).
```
Test case added.

All other nits addressed

mgartner reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> 

> Reviewed 6 of 27 files at r2, 20 of 20 files at r4, 18 of 19 files at r5, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/89884#-:-NFU8Yft9ln61HTLhr7X:bg7igkv)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @jordanlewis, @miretskiy, and @yuzefovich)



<!-- Sent from Reviewable.io -->

yuzefovich reviewed (commented):
Very impressive numbers! Nice work! I didn't review too carefully, but I still got some nits for you :)

> Reviewed 1 of 17 files at r1, 6 of 27 files at r2, 13 of 20 files at r4, 11 of 19 files at r5, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/89884#-:-NFUcMre48KF8ITpvgzZ:btkp3zx)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @jordanlewis, @mgartner, and @miretskiy)

___
*[`pkg/util/json/config.go` line 32 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/89884#-NFUWlrs8wQJxMT2ILz_:-NFUWlrs8wQJxMT2ILza:bgxdl6h) ([raw file](https://github.com/cockroachdb/cockroach/blob/c681733cd1a8c0e2f4539cee89abd6e29e231932/pkg/util/json/config.go#L32)):*
> ```Go
> 
> // WithGoStandardParser returns an option that forces the use of go parser
> // (encoding/json)
> ```

nit: missing period.
___
*[`pkg/util/json/parser.go` line 206 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/89884#-NFUY7sE3rNr55WSOeI1:-NFUY7sE3rNr55WSOeI2:bpptsyf) ([raw file](https://github.com/cockroachdb/cockroach/blob/c681733cd1a8c0e2f4539cee89abd6e29e231932/pkg/util/json/parser.go#L206)):*
> ```Go
> var errUnexpectedState = errors.New("unexpected state machine state")
> 
> // pushArray adds array builder and transition state to
> ```

nit: `s/transition/transitions/`.
___
*[`pkg/util/json/parser.go` line 219 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/89884#-NFUYExhB0HAUWwjFxWk:-NFUYExhB0HAUWwjFxWl:b-u7nnft) ([raw file](https://github.com/cockroachdb/cockroach/blob/c681733cd1a8c0e2f4539cee89abd6e29e231932/pkg/util/json/parser.go#L219)):*
> ```Go
> }
> 
> // buildArray builds top array value, and adjust stack appropriately.
> ```

nit: `s/adjust/adjusts/`.
___
*[`pkg/util/json/parser.go` line 274 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/89884#-NFUYj-G7Dzqjavyd_oS:-NFUYj-G7Dzqjavyd_oT:bpptsyf) ([raw file](https://github.com/cockroachdb/cockroach/blob/c681733cd1a8c0e2f4539cee89abd6e29e231932/pkg/util/json/parser.go#L274)):*
> ```Go
> 
> // stackReturn returns json object to the top of the stack
> // and transition state machine to the next state.
> ```

nit: `s/transition/transitions/`.
___
*[`pkg/util/json/tokenizer/buffer.go` line 16 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/89884#-NFUZtrZD8GVMPE1fSx8:-NFUZtrZD8GVMPE1fSx9:b10x0z) ([raw file](https://github.com/cockroachdb/cockroach/blob/c681733cd1a8c0e2f4539cee89abd6e29e231932/pkg/util/json/tokenizer/buffer.go#L16)):*
> ```Go
> 
> // buffer, is very much like bytes.Buffer, but is simpler. In particular,
> // because we don't need to support entirety bytes.Buffer interface, the
> ```

nit: `s/entirety/entirety of/`.
___
*[`pkg/util/json/tokenizer/buffer.go` line 17 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/89884#-NFUZyi0DxfT9jTPj-jB:-NFUZyi0DxfT9jTPj-jC:bdp9mhe) ([raw file](https://github.com/cockroachdb/cockroach/blob/c681733cd1a8c0e2f4539cee89abd6e29e231932/pkg/util/json/tokenizer/buffer.go#L17)):*
> ```Go
> // buffer, is very much like bytes.Buffer, but is simpler. In particular,
> // because we don't need to support entirety bytes.Buffer interface, the
> // functions (s.a. AppendByte) are simpler, and as a result can be inlined.
> ```

We have a GCAssert linter to ensure things are inlined. You'd need to add `//gcassert:inline` comment right before the relevant functions and then add this package into `TestGCAssert` linter if you'd like that. (Note that at the moment the linter is skipped, but I'm working on un-skipping it.)
___
*[`pkg/util/json/tokenizer/buffer.go` line 83 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/89884#-NFU_ns56P4s3zEYKB4m:-NFU_ns56P4s3zEYKB4n:bgxdl6h) ([raw file](https://github.com/cockroachdb/cockroach/blob/c681733cd1a8c0e2f4539cee89abd6e29e231932/pkg/util/json/tokenizer/buffer.go#L83)):*
> ```Go
> 	}
> 
> 	// Cribbed from bytes.Buffer
> ```

nit: missing period.
___
*[`pkg/util/json/tokenizer/buffer.go` line 92 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/89884#-NFUaGetCEXUiyjclrIg:-NFUaGetCEXUiyjclrIh:be4p7iv) ([raw file](https://github.com/cockroachdb/cockroach/blob/c681733cd1a8c0e2f4539cee89abd6e29e231932/pkg/util/json/tokenizer/buffer.go#L92)):*
> ```Go
> 	copy(nb, buf)
> 	// Make whole cap available since append may produce larger slice.
> 	return nb[:cap(nb)]
> ```

The fact that we have `len(buffer.buf) == cap(buffer.buf)` seems non-trivial, so I would explicitly also mention this in the comment to `buffer` struct.
___
*[`pkg/util/json/tokenizer/decoder.go` line 51 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/89884#-NFUag9OADyUbz-aZ4yB:-NFUag9OADyUbz-aZ4yC:bgxdl6h) ([raw file](https://github.com/cockroachdb/cockroach/blob/c681733cd1a8c0e2f4539cee89abd6e29e231932/pkg/util/json/tokenizer/decoder.go#L51)):*
> ```Go
> }
> 
> // MakeDecoder returns decoder for the input data
> ```

nit: missing period.
___
*[`pkg/util/json/tokenizer/scanner.go` line 88 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/89884#-NFUVQz05R2yawv9yl4q:-NFUVQz05R2yawv9yl4r:b-be2md) ([raw file](https://github.com/cockroachdb/cockroach/blob/bfd8b1d1d751d607f9ff0cfe90817a5c1af90794/pkg/util/json/tokenizer/scanner.go#L88)):*
> ```Go
> }
> 
> // Next returns a []byte referencing the the next lexical token in the stream.
> ```

nit: `s/the the/the/`.
___
*[`pkg/util/json/tokenizer/scanner.go` line 208 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/89884#-NFUbC7A5D-6NI5BF-H7:-NFUbC7A5D-6NI5BF-H8:bgxdl6h) ([raw file](https://github.com/cockroachdb/cockroach/blob/c681733cd1a8c0e2f4539cee89abd6e29e231932/pkg/util/json/tokenizer/scanner.go#L208)):*
> ```Go
> func (s *Scanner) parseStringSlow(pos int) ([]byte, int) {
> 	w := s.buf()
> 	// Sanity check
> ```

nit: missing period.


<!-- Sent from Reviewable.io -->

miretskiy commented:
> Very impressive numbers! Nice work! I didn't review too carefully, but I still got some nits for you :)

Thanks; nits addressed.

> We have a GCAssert linter to ensure things are inlined
I don't know if this code qualifies for these kinds of protections.  I don't like to go on "inline hunting"

miretskiy commented:
bors r+
